src/builtins/ecdsa/ecdsa.ts

```diff
@@ -0,0 +1,39 @@
+function sign_message(message: bigint, private_key: bigint): SignatureType {
+  var k: bigint = getRandomBigInt(ORDER - BigInt(1)) + BigInt(1);
+  var r_point: AffinePoint = ec_mul(k, GENERATOR);
+  var r: bigint = r_point.x % ORDER;
+  if (r === BigInt(0)) {
+    return sign_message(message, private_key);
+  }
+  var s: bigint = modMul(
+    modInv(k, ORDER),
+    modAdd(message, modMul(r, private_key, ORDER), ORDER),
+    ORDER,
+  );
+  return { r: r, s: s } as SignatureType;
+}
+
+function verify_sig(
+  signature: SignatureType,
+  message: bigint,
+  public_key: AffinePoint,
+) {
+  var s_inv: bigint = modInv(signature.s, ORDER);
+  var u: bigint = modMul(message, s_inv, ORDER);
+  var v: bigint = modMul(signature.r, s_inv, ORDER);
+  var c: AffinePoint = ec_add(ec_mul(u, GENERATOR), ec_mul(v, public_key));
+  return c.x === signature.r;
+}
+
+function getRandomBigInt(n: bigint): bigint {
```

Comment:
I'd go for something string based instead and avoid the "random" for loop, similar to https://stackoverflow.com/a/49232990/4444546

```
target = n.toString(16).length
// generate random hex string with the appropriate length
return BitIng(random) % n
```

Reply:
Tell me if you had smth else in mind